mobile/native/src/db/models.rs

```diff
@@ -442,58 +455,36 @@ impl Position {
         Ok(position)
     }
 
-    // sets the position to rollover and updates the new expiry timestamp.
-    pub fn rollover(
-        conn: &mut SqliteConnection,
-        contract_symbol: ContractSymbol,
-        expiry_timestamp: OffsetDateTime,
-    ) -> Result<()> {
+    /// Update the position after the rollover protocol has started.
+    pub fn start_rollover(conn: &mut SqliteConnection, updated_position: Position) -> Result<()> {
         let affected_rows = diesel::update(positions::table)
-            .filter(schema::positions::contract_symbol.eq(contract_symbol))
-            .set((
-                positions::expiry_timestamp.eq(expiry_timestamp.unix_timestamp()),
-                positions::state.eq(PositionState::Rollover),
-                positions::updated_timestamp.eq(OffsetDateTime::now_utc().unix_timestamp()),
-            ))
+            .set(updated_position)
             .execute(conn)?;
 
-        ensure!(affected_rows > 0, "Could not set position to rollover");
+        if affected_rows == 0 {
+            bail!("Could not start rollover in DB");
+        }
 
         Ok(())
     }
 
-    /// Updates the status of the given order in the DB.
-    pub fn update_position(conn: &mut SqliteConnection, position: Position) -> Result<()> {
-        let Position {
-            contract_symbol,
-            leverage,
-            quantity,
-            direction,
-            average_entry_price,
-            liquidation_price,
-            state,
-            collateral,
-            creation_timestamp: _,
-            expiry_timestamp,
-            updated_timestamp,
-            order_matching_fees,
-            ..
-        } = position;
+    /// Update the position after the rollover protocol has ended.
+    pub fn finish_rollover(conn: &mut SqliteConnection, updated_position: Position) -> Result<()> {
+        let affected_rows = diesel::update(positions::table)
+            .filter(positions::state.eq(PositionState::Rollover))
```

Comment:
If we always assume that there is ever only one position, we might want to consider removing this filter as it would potentially only lead to errors if the position is not in the state we assume it should be.

Reply:
That makes some sense. Although in theory we could be catching an actual error here.

Reply:
I opened an issue because I think your comment applies to other spots too: https://github.com/get10101/10101/issues/2599.